tests/lib/nm/dns_test.py

```diff
@@ -0,0 +1,130 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+import pytest
+
+from libnmstate import metadata
+from libnmstate.schema import DNS
+import libnmstate.nm.connection as nm_connection
+import libnmstate.nm.ipv4 as nm_ipv4
+import libnmstate.nm.ipv6 as nm_ipv6
+
+
+NM_USE_DEFAULT_DNS_PRIORITY = 0
+
+
+def _get_test_dns_v4():
+    return {
+        metadata.DNS_PRIORITY: 40,
+        DNS.SERVER: ['8.8.8.8', '1.1.1.1'],
+        DNS.SEARCH: ['example.org', 'example.com']
+    }
+
+
+def _get_test_dns_v6():
+    return {
+        metadata.DNS_PRIORITY: 40,
+        DNS.SERVER: ['2001:4860:4860::8888', '2606:4700:4700::1111'],
+        DNS.SEARCH: ['example.net', 'example.edu']
+    }
+
+
+parametrize_ip_ver = pytest.mark.parametrize(
+    'nm_ip',
+    [(nm_ipv4), (nm_ipv6)],
+    ids=['ipv4', 'ipv6'])
+
+
+parametrize_ip_ver_dns = pytest.mark.parametrize(
+    'nm_ip, get_test_dns_func',
+    [(nm_ipv4, _get_test_dns_v4),
+     (nm_ipv6, _get_test_dns_v6)],
+    ids=['ipv4', 'ipv6'])
+
+
+@parametrize_ip_ver
+def test_add_dns_empty(nm_ip):
+    dns_conf = {}
+    setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: dns_conf
+    }, base_con_profile=None)
+
+    _assert_dns(setting_ip, dns_conf)
+
+
+@parametrize_ip_ver_dns
+def test_add_dns(nm_ip, get_test_dns_func):
+    dns_conf = get_test_dns_func()
+    setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: dns_conf
+    }, base_con_profile=None)
+
+    _assert_dns(setting_ip, dns_conf)
+
+
+@parametrize_ip_ver_dns
+def test_add_dns_duplicate_server(nm_ip, get_test_dns_func):
+    dns_conf = get_test_dns_func()
+    dns_conf[DNS.SERVER] = [dns_conf[DNS.SERVER][0], dns_conf[DNS.SERVER][0]]
+    setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: dns_conf
+    }, base_con_profile=None)
+
+    dns_conf[DNS.SERVER] = [dns_conf[DNS.SERVER][0]]
+    _assert_dns(setting_ip, dns_conf)
+
+
+@parametrize_ip_ver_dns
+def test_add_dns_duplicate_search(nm_ip, get_test_dns_func):
+    dns_conf = get_test_dns_func()
+    dns_conf[DNS.SEARCH] = [dns_conf[DNS.SEARCH][0], dns_conf[DNS.SEARCH][0]]
+    setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: dns_conf
+    }, base_con_profile=None)
+
+    dns_conf[DNS.SEARCH] = [dns_conf[DNS.SEARCH][0]]
+    _assert_dns(setting_ip, dns_conf)
+
+
+@parametrize_ip_ver_dns
+def test_clear_dns(nm_ip, get_test_dns_func):
+    dns_conf = get_test_dns_func()
+    setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: dns_conf
+    }, base_con_profile=None)
+    con_profile = nm_connection.ConnectionProfile()
+    con_profile.create([setting_ip])
+    new_setting_ip = nm_ip.create_setting({
+        'enabled': True,
+        metadata.DNS_METADATA: {}
+    }, base_con_profile=con_profile.profile)
+
+    _assert_dns(new_setting_ip, {})
```

Comment:
I think you need to show that the original `setting_ip` has its dns cleared.
The `new_setting_ip` should be the (`is`) the `setting_ip`.